### ISSUE TITLE
Form validation and completion modal

### ISSUE THREAD

**Issue by ojeytonwilliams** [freeCodeCamp/freeCodeCamp#57518]
The logic that determines if a challenge shows a completion modal on success is spread across the codebase. It's particularly complex for the front and backend projects, since it depends on validation details. It's possible to run tests with unsubmittable values (i.e. local urls), but then the solution modal does not appear.

Instead, can we decouple form validation from the logic that controls the modal? In principle we can replace the validation with tests, so that the learner can, say, use a local server for testing. They'd see a failing test, but it would be obvious how to fix it: host the app remotely.

If we do this we can make the modal's appearance always depend on two things: 1) challenge type 2) passing tests.

@ShaunSHamilton I know you've worked on this in the past: any thoughts?

**Comment by ShaunSHamilton** [comment]
This sounds fine, except that means adding many of the same tests to hundreds of challenges, plus some challenges (e.g. Front-End Libraries Certification Projects) do not have any tests.

The only way I can see to _slightly_ mitigate this is to use the MDX imports, which were working a few years ago.

**Comment by ojeytonwilliams** [comment]
If it's only a few hundred challenges, that doesn't sound too bad. It should just be a matter of copying and pasting after all.

**Comment by ojeytonwilliams** [comment]
As far I'm concerned, in this issue, what matters is that form validation should only affect _if_ the tests can be run, not the outcome. Both types of validation (errors and warnings) should be totally fine as long as they respect that condition.

**Comment by Rajatsharma003** [comment]
The logic for determining whether a challenge displays a completion modal upon success is currently distributed across the codebase, making it particularly challenging for both frontend and backend projects. This complexity arises because the modal's behavior is tightly coupled with form validation details.

_For instance, it's possible to run tests with unsubmittable values (e.g., local URLs), but the solution modal does not appear, leading to confusion for the user._

**Proposed Solution:**
We should consider decoupling form validation from the logic controlling the modal. Instead of relying on validation, we can base the modal's appearance purely on:

_The type of challenge._
Whether all tests have passed.
This approach would allow learners to test their solutions using a local server. They might encounter failing tests (e.g., due to a local URL), but it would be clear that the issue lies with hosting. Learners would then know they need to host their app remotely to pass the challenge.

By implementing this change, we could simplify the logic and improve the user experience.

@ShaunSHamilton Since you’ve worked on this in the past, I’d appreciate your thoughts on this approach.

**Comment by ilenia-magoni** [comment]
> This approach would allow learners to test their solutions using a local server. They might encounter failing tests (e.g., due to a local URL), but it would be clear that the issue lies with hosting. Learners would then know they need to host their app remotely to pass the challenge.

this is the current behaviour, you can test with `localhost`, but there is an alert on the input element that you need to use an app hosted somewhere

**Comment by ojeytonwilliams** [comment]
That's right, but the problem is that it doesn't just show the alert, it also prevents the completion modal from appearing.

**Comment by ilenia-magoni** [comment]
Doesn't that happens only for certification projects where the link to a live app is mandatory?

Edit: right, we used those projects as labs now, so the live app link should not be mandatory anymore as we should not be saving those... new challengeType? 😓
